{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/shankaracharya/machine_learning/blob/master/colabs/visualization_modality.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "y4k5PhBwckLM"
      },
      "source": [
        "# Visualizing predictions\n",
        "\n",
        "AlphaGenome can make predictions for various different modalities. We built a visualization library to aid interpretation of the outputs from the model. In this tutorial notebook you will:\n",
        "*   Learn how to visualize different data modalities for a specific genomic interval (and any variants) from model output.\n",
        "*   Understand how to use the primary functionality of the visualization library `alphagenome.visualization`.\n",
        "\n",
        "We also recommend visiting [\"Visualization Basics\"](https://www.alphagenomedocs.com/visualization_library_basics.html) for an overview of the library and its primary functionality."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HWAeLvV6Nio8"
      },
      "source": [
        "```{tip}\n",
        "Open this tutorial in Google Colab for interactive viewing.\n",
        "```"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "e1dQWw1HJap7"
      },
      "outputs": [],
      "source": [
        "# @title Install AlphaGenome\n",
        "\n",
        "# @markdown Run this cell to install AlphaGenome.\n",
        "from IPython.display import clear_output\n",
        "! pip install alphagenome\n",
        "clear_output()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PhxWI_U1KbdK"
      },
      "source": [
        "## Setup and imports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QX39iOcbKbdK"
      },
      "outputs": [],
      "source": [
        "from alphagenome import colab_utils\n",
        "from alphagenome.data import gene_annotation, genome, track_data, transcript\n",
        "from alphagenome.models import dna_client\n",
        "from alphagenome.visualization import plot_components\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "import pandas as pd"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WuAGnIXs5-ER"
      },
      "source": [
        "### Load model and auxiliary objects"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KUa_l1ceDmmz"
      },
      "outputs": [],
      "source": [
        "dna_model = dna_client.create(colab_utils.get_api_key())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Rg_upvGia3-j"
      },
      "outputs": [],
      "source": [
        "# Load metadata objects for human.\n",
        "output_metadata = dna_model.output_metadata(\n",
        "    organism=dna_client.Organism.HOMO_SAPIENS\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "XxkkkStMuJn5"
      },
      "outputs": [],
      "source": [
        "# Load gene annotations (from GENCODE).\n",
        "gtf = pd.read_feather(\n",
        "    'https://storage.googleapis.com/alphagenome/reference/gencode/'\n",
        "    'hg38/gencode.v46.annotation.gtf.gz.feather'\n",
        ")\n",
        "\n",
        "# Filter to protein-coding genes and highly supported transcripts.\n",
        "gtf_transcript = gene_annotation.filter_transcript_support_level(\n",
        "    gene_annotation.filter_protein_coding(gtf), ['1']\n",
        ")\n",
        "\n",
        "# Extractor for identifying transcripts in a region.\n",
        "transcript_extractor = transcript.TranscriptExtractor(gtf_transcript)\n",
        "\n",
        "# Also define an extractor that fetches only the longest transcript per gene.\n",
        "gtf_longest_transcript = gene_annotation.filter_to_longest_transcript(\n",
        "    gtf_transcript\n",
        ")\n",
        "longest_transcript_extractor = transcript.TranscriptExtractor(\n",
        "    gtf_longest_transcript\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "O5jsAYA8ak3T"
      },
      "source": [
        "## Gene expression"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "E41aNDzmd02o"
      },
      "source": [
        "Gene expression is primarily captured by the model outputs `RNA_SEQ` and `CAGE`.\n",
        "Here is an example of expression predictions for colon tissue in a genomic\n",
        "interval containing the gene *APOL4* . Positions of the longest transcript per\n",
        "gene for genes in this interval are shown at the top:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8DF1fR_DtLQ2"
      },
      "outputs": [],
      "source": [
        "# Define interval to make predictions for (used throughout this tutorial).\n",
        "# Note that the interval width must be one of the supported sequence lengths.\n",
        "interval = genome.Interval('chr22', 36_150_498, 36_252_898).resize(\n",
        "    dna_client.SEQUENCE_LENGTH_1MB\n",
        ")\n",
        "\n",
        "# Define the tissues/cell-types to predict expression for.\n",
        "ontology_terms = [\n",
        "    'UBERON:0001159',  # Colon - Sigmoid.\n",
        "    'UBERON:0001155',  # Colon - Transverse.\n",
        "]\n",
        "\n",
        "# Make predictions.\n",
        "output = dna_model.predict_interval(\n",
        "    interval=interval,\n",
        "    requested_outputs={\n",
        "        dna_client.OutputType.RNA_SEQ,\n",
        "        dna_client.OutputType.CAGE,\n",
        "    },\n",
        "    ontology_terms=ontology_terms,\n",
        ")\n",
        "\n",
        "# Extract the longest transcripts per gene for this interval.\n",
        "longest_transcripts = longest_transcript_extractor.extract(interval)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7Ns0fN4eZ2Zb"
      },
      "source": [
        "Build the plot by constructing multiple plotting panels or 'components' and\n",
        "feeding these into the `plot_components.plot()` function. Refer to\n",
        "[visualization basics guide](https://www.alphagenomedocs.com/visualization_library_basics.html#visualization-basics)\n",
        "for a description of available plotting components."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Gqe0PA-2DLn2"
      },
      "outputs": [],
      "source": [
        "# Build plot.\n",
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output.rna_seq,\n",
        "            ylabel_template='RNA_SEQ: {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output.cage,\n",
        "            ylabel_template='CAGE: {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "    ],\n",
        "    interval=interval,\n",
        "    title='Predicted RNA Expression (RNA_SEQ, CAGE) for colon tissue',\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AVF7aiwI8eh7"
      },
      "source": [
        "Note that the positive and negative stranded tracks have clearly different\n",
        "signals. Let's check which genes are transcribed from the positive strand:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "M-nGhwYw7J1h"
      },
      "outputs": [],
      "source": [
        "[t.info['gene_name'] for t in longest_transcripts if t.strand == '+']"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-0rVlCL-8lSF"
      },
      "source": [
        "For both of the positive stranded `RNA_SEQ` tracks, we can see that RNA-seq\n",
        "levels are highest around *APOL1*. Similarly for `CAGE`, the peaks on the\n",
        "positive strand fall around the start of this gene. We do not see peaks around\n",
        "*APOL5* as this gene is not expressed in colon tissue\n",
        "[(GTEx)](https://www.gtexportal.org/home/gene/APOL5#geneExpression)."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lxOBhD5vw4eT"
      },
      "source": [
        "### Visualize the effect of a variant\n",
        "\n",
        "A previously identified variant in this region (`chr22_36201698_A_C`) affects\n",
        "both the expression and splicing of the *APOL4* gene. Specifically, the\n",
        "alternative allele (C) is linked to reduced *APOL4* expression.\n",
        "\n",
        "To visualize what AlphaGenome predicts for this variant, we can:\n",
        "\n",
        "*   Compute predictions for the REF and ALT sequences using\n",
        "    `dna_model.predict_variant`.\n",
        "*   Remove positive-stranded tracks (as *APOL4* is transcribed from the negative\n",
        "    DNA strand).\n",
        "*   Zoom in to the region around the gene *APOL4*.\n",
        "*   Highlight the location of the variant using\n",
        "    `plot_components.VariantAnnotation`.\n",
        "*   Increase the relative height of the transcripts section to better view the\n",
        "    gene structure as annotated by GENCODE."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ORkUrqj3xAGH"
      },
      "outputs": [],
      "source": [
        "# Define the variant of interest.\n",
        "variant_string = 'chr22:36201698:A>C'\n",
        "variant = genome.Variant.from_str(variant_string)\n",
        "variant"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "gDb89LLPKGoh"
      },
      "outputs": [],
      "source": [
        "# Make predictions for sequences containing the REF and ALT alleles.\n",
        "output = dna_model.predict_variant(\n",
        "    interval=interval,\n",
        "    variant=variant,\n",
        "    requested_outputs={\n",
        "        dna_client.OutputType.RNA_SEQ,\n",
        "        dna_client.OutputType.CAGE,\n",
        "    },\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "x8Vp0rT55NhT"
      },
      "outputs": [],
      "source": [
        "# Zoom in on the region around APOL4.\n",
        "apol4_interval = gene_annotation.get_gene_interval(gtf, gene_symbol='APOL4')\n",
        "\n",
        "# Add 1KB on either side of the gene body.\n",
        "apol4_interval.resize_inplace(apol4_interval.width + 1000)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5s7qC5Nvw2yb"
      },
      "outputs": [],
      "source": [
        "# Define the colors for REF and ALT predictions.\n",
        "ref_alt_colors = {'REF': 'dimgrey', 'ALT': 'red'}\n",
        "\n",
        "# Build plot.\n",
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        # RNA-seq tracks.\n",
        "        plot_components.OverlaidTracks(\n",
        "            tdata={\n",
        "                'REF': output.reference.rna_seq.filter_to_nonpositive_strand(),\n",
        "                'ALT': output.alternate.rna_seq.filter_to_nonpositive_strand(),\n",
        "            },\n",
        "            colors=ref_alt_colors,\n",
        "            ylabel_template='{biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "        # CAGE track.\n",
        "        plot_components.OverlaidTracks(\n",
        "            tdata={\n",
        "                'REF': output.reference.cage.filter_to_nonpositive_strand(),\n",
        "                'ALT': output.alternate.cage.filter_to_nonpositive_strand(),\n",
        "            },\n",
        "            colors=ref_alt_colors,\n",
        "            ylabel_template='{biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "    ],\n",
        "    annotations=[plot_components.VariantAnnotation([variant])],\n",
        "    interval=apol4_interval,\n",
        "    title='Effect of variant on predicted RNA Expression in colon tissue',\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6hvjr9-LeKNg"
      },
      "source": [
        "You can see here that the predicted RNA-seq values for the alternative allele\n",
        "are much lower across the gene (ALT lines are lower than REF lines), which is\n",
        "also reflected in the bottom CAGE track. Additionally, you can see where the\n",
        "variant might be inducing an exon skipping event (in the exon where the ALT line\n",
        "is zero but the REF has a peak)."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GEasjDBTbvwB"
      },
      "source": [
        "### Plot custom annotation (e.g., polyadenylation sites)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Oa_fgTCJbtFJ"
      },
      "outputs": [],
      "source": [
        "# Run predictions with an additional ontology to highlight differences across\n",
        "# tissues.\n",
        "ontology_terms = [\n",
        "    'UBERON:0001159',  # Colon - Sigmoid.\n",
        "    'UBERON:0002048',  # lung.\n",
        "]\n",
        "\n",
        "# Make predictions.\n",
        "output = dna_model.predict_interval(\n",
        "    interval=interval,\n",
        "    requested_outputs={\n",
        "        dna_client.OutputType.RNA_SEQ,\n",
        "    },\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "K2fDMPmnLJNb"
      },
      "source": [
        "There are 3 annotated pA sites overlapping APOL4 transcripts, as defined by\n",
        "[PolyADBv3](https://exon.apps.wistar.org/polya_db/v3/misc/download.php) [(Wang et al. 2018)](https://pubmed.ncbi.nlm.nih.gov/29069441/).\n",
        "We can define these locations as 1bp intervals in hg38 coordinates."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Qxv7hhJrcwy6"
      },
      "outputs": [],
      "source": [
        "# Define pA sites in hg38 coordinates.\n",
        "apol4_pAs = [genome.Interval('chr22', 36_189_128, 36_189_129, '-'),\n",
        "             genome.Interval('chr22', 36_190_089, 36_190_090, '-'),\n",
        "             genome.Interval('chr22', 36_190_144, 36_190_145, '-')]\n",
        "\n",
        "# Define plotting interval as first/last pA plus some offset distance.\n",
        "\n",
        "offset = 200\n",
        "pA_interval = genome.Interval('chr22',\n",
        "                              36_189_128 - offset,\n",
        "                              36_190_145 + offset,\n",
        "                              '-')\n",
        "\n",
        "# Define intervals annotation.\n",
        "pA_labels = ['pA_3', 'pA_2', 'pA_1']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JX1wMyFrdZ4x"
      },
      "outputs": [],
      "source": [
        "# Build plot.\n",
        "# NOTE: Depending on annotation distance and interval zoom some annotations may # appear as overlapping.\n",
        "\n",
        "plot = plot_components.plot(\n",
        "    [plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        # RNA-seq tracks.\n",
        "        plot_components.Tracks(\n",
        "            tdata = output.rna_seq.filter_to_negative_strand(),\n",
        "            ylabel_template = 'RNA_SEQ: {biosample_name} ({strand})\\n{name}',\n",
        "            shared_y_scale = True,\n",
        "            )\n",
        "    ],\n",
        "    annotations=[\n",
        "        plot_components.IntervalAnnotation(apol4_pAs,\n",
        "                                           alpha = 1,\n",
        "                                           labels = pA_labels,\n",
        "                                           label_angle = 90)],\n",
        "    interval = pA_interval,\n",
        "    title='APOL4 polyadenylation sites annotation',\n",
        "\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "q0lfa_JphWux"
      },
      "source": [
        "## Chromatin accessibility"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iMBTQtrvmUQ7"
      },
      "source": [
        "Chromatin accessibility is primarily captured by the model outputs `DNASE` and\n",
        "`ATAC`. Here is an example of predictions for the same genomic interval\n",
        "containing the gene *APOL4*. There are a large variety of tissues and cell-types\n",
        "available, but let's focus on plotting predictions for intestinal tract tissues:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "MfwL6sNJCtgZ"
      },
      "outputs": [],
      "source": [
        "# List of IDs corresponding to various intestinal tissues.\n",
        "ontology_terms = [\n",
        "    'UBERON:0000317',\n",
        "    'UBERON:0001155',\n",
        "    'UBERON:0001157',\n",
        "    'UBERON:0001159',\n",
        "    'UBERON:0004992',\n",
        "    'UBERON:0008971',\n",
        "]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1E196OgliWf1"
      },
      "outputs": [],
      "source": [
        "# Make predictions.\n",
        "output = dna_model.predict_interval(\n",
        "    interval,\n",
        "    requested_outputs={\n",
        "        dna_client.OutputType.DNASE,\n",
        "        dna_client.OutputType.ATAC,\n",
        "    },\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nWv31y-IzNro"
      },
      "source": [
        "Chromatin accessibility can indicate regions of regulatory activity. Let's use\n",
        "`plot_components.IntervalAnnotation()` to plot additional annotation on putative\n",
        "promoter regions from GENCODE\n",
        "[human](https://www.gencodegenes.org/human/release_46.html) or [mouse](https://www.gencodegenes.org/mouse/release_M23.html)\n",
        "to see if they overlap regions predicted to be accessible:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bnIspPV8wkQ9"
      },
      "outputs": [],
      "source": [
        "# Two putative promoter intervals from Ensembl.\n",
        "promoter_intervals = [\n",
        "    genome.Interval(\n",
        "        'chr22', 36_201_799, 36_202_681, name='Ensembl_promoter:ENSR00001367790'\n",
        "    ),\n",
        "    genome.Interval(\n",
        "        'chr22', 36_204_705, 36_205_330, name='Ensembl_promoter:ENSR00001367792'\n",
        "    ),\n",
        "]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "S6mXwvbAimQ-"
      },
      "outputs": [],
      "source": [
        "# Build plot.\n",
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output.dnase,\n",
        "            ylabel_template='DNASE: {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output.atac,\n",
        "            ylabel_template='ATAC: {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "    ],\n",
        "    # Plot an 8kb window around the variant.\n",
        "    interval=variant.reference_interval.resize(8000),\n",
        "    annotations=[\n",
        "        plot_components.VariantAnnotation([variant]),\n",
        "        plot_components.IntervalAnnotation(promoter_intervals),\n",
        "    ],\n",
        "    title='Predicted chromatin accessibility (DNASE, ATAC) for colon tissue',\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MXxYBvMlvGjG"
      },
      "source": [
        "Some observations from this plot:\n",
        "*   Elevated `DNASE` and `ATAC` signals overlap\n",
        "both putative promoter regions highlighted in grey, but additional chromatin\n",
        "accessibility peaks suggest there may be other regulatory elements in this\n",
        "region.\n",
        "*   `DNASE` and `ATAC` signals are similar but not identical, reflecting\n",
        "differences in assay protocol.\n",
        "*  Accessibility is not especially high around the variant (orange line)."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ibR_Luiqzwve"
      },
      "source": [
        "## Splicing"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cASBC9c0zwve"
      },
      "source": [
        "Splicing effects are primarily captured by the model outputs `SPLICE_SITES`,\n",
        "`SPLICE_SITE_USAGE`, and `SPLICE_JUNCTIONS`. Here is an example of splicing\n",
        "predictions in a genomic interval containing the gene *APOL4*.\n",
        "\n",
        "Since `RNA_SEQ` data also captures splicing patterns, we can plot it together\n",
        "with the splicing outputs, allowing us to see how predictions from different\n",
        "modalities are related:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bzgBbXQqzwve"
      },
      "outputs": [],
      "source": [
        "# List of IDs corresponding to various intestinal tissues.\n",
        "ontology_terms = [\n",
        "    'UBERON:0001157',\n",
        "    'UBERON:0001159',\n",
        "]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PDDdUrCgzwve"
      },
      "outputs": [],
      "source": [
        "# Make predictions for splicing outputs and RNA_SEQ.\n",
        "output = dna_model.predict_interval(\n",
        "    interval=interval,\n",
        "    requested_outputs={\n",
        "        dna_client.OutputType.RNA_SEQ,\n",
        "        dna_client.OutputType.SPLICE_SITES,\n",
        "        dna_client.OutputType.SPLICE_SITE_USAGE,\n",
        "        dna_client.OutputType.SPLICE_JUNCTIONS,\n",
        "    },\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JAcrIt_-_ENZ"
      },
      "source": [
        "Note that `SPLICE_SITES` are tissue-agnostic predictions, so the\n",
        "`ontology_terms` filter is not applied to it."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4JUQL4jWPw66"
      },
      "outputs": [],
      "source": [
        "output.splice_sites.metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3p4XZwL76F6X"
      },
      "outputs": [],
      "source": [
        "# Build plot.\n",
        "# Since APOL4 is on the negative DNA strand, we use `filter_negative_strand` to\n",
        "# consider only negative stranded splice predictions.\n",
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output.splice_sites.filter_to_negative_strand(),\n",
        "            ylabel_template='SPLICE SITES: {name} ({strand})',\n",
        "        ),\n",
        "    ],\n",
        "    interval=apol4_interval,\n",
        "    title='Predicted splicing effects for colon tissue',\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pUkCJQF68le0"
      },
      "source": [
        "Note how the model predicts acceptor splice sites near the beginnings of exons\n",
        "and donor splice sites near the ends of exons, whereas splice site usage\n",
        "predictions have peaks on both sides of exons.\n",
        "\n",
        "We can visualize the junction split read predictions as arcs using\n",
        "`plot_components.Sashimi()`. We can also visualize the effect of the variant by\n",
        "showing both `RNA_SEQ` and splicing predictions on the same plot:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZsW6gqZ3zwve"
      },
      "outputs": [],
      "source": [
        "# Make predictions for REF and ALT alleles.\n",
        "output = dna_model.predict_variant(\n",
        "    interval=interval,\n",
        "    variant=variant,\n",
        "    requested_outputs={\n",
        "        dna_client.OutputType.RNA_SEQ,\n",
        "        dna_client.OutputType.SPLICE_SITES,\n",
        "        dna_client.OutputType.SPLICE_SITE_USAGE,\n",
        "        dna_client.OutputType.SPLICE_JUNCTIONS,\n",
        "    },\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ou8Yju8s-I0R"
      },
      "outputs": [],
      "source": [
        "# Get all transcripts, not just the longest one per gene.\n",
        "transcripts = transcript_extractor.extract(interval)\n",
        "\n",
        "ref_output = output.reference\n",
        "alt_output = output.alternate\n",
        "\n",
        "# Build plot.\n",
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(transcripts),\n",
        "        plot_components.Sashimi(\n",
        "            ref_output.splice_junctions\n",
        "            .filter_to_strand('-')\n",
        "            .filter_by_tissue('Colon_Transverse'),\n",
        "            ylabel_template='Reference {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "        plot_components.Sashimi(\n",
        "            alt_output.splice_junctions\n",
        "            .filter_to_strand('-')\n",
        "            .filter_by_tissue('Colon_Transverse'),\n",
        "            ylabel_template='Alternate {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "        plot_components.OverlaidTracks(\n",
        "            tdata={\n",
        "                'REF': ref_output.rna_seq.filter_to_nonpositive_strand(),\n",
        "                'ALT': alt_output.rna_seq.filter_to_nonpositive_strand(),\n",
        "            },\n",
        "            colors=ref_alt_colors,\n",
        "            ylabel_template='RNA_SEQ: {biosample_name} ({strand})\\n{name}',\n",
        "        ),\n",
        "        plot_components.OverlaidTracks(\n",
        "            tdata={\n",
        "                'REF': ref_output.splice_sites.filter_to_nonpositive_strand(),\n",
        "                'ALT': alt_output.splice_sites.filter_to_nonpositive_strand(),\n",
        "            },\n",
        "            colors=ref_alt_colors,\n",
        "            ylabel_template='SPLICE SITES: {name} ({strand})',\n",
        "        ),\n",
        "        plot_components.OverlaidTracks(\n",
        "            tdata={\n",
        "                'REF': (\n",
        "                    ref_output.splice_site_usage.filter_to_nonpositive_strand()\n",
        "                ),\n",
        "                'ALT': (\n",
        "                    alt_output.splice_site_usage.filter_to_nonpositive_strand()\n",
        "                ),\n",
        "            },\n",
        "            colors=ref_alt_colors,\n",
        "            ylabel_template=(\n",
        "                'SPLICE SITE USAGE: {biosample_name} ({strand})\\n{name}'\n",
        "            ),\n",
        "        ),\n",
        "    ],\n",
        "    interval=apol4_interval,\n",
        "    annotations=[plot_components.VariantAnnotation([variant])],\n",
        "    title='Predicted REF vs. ALT effects of variant in colon tissue',\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wYXc0CmVqyVD"
      },
      "source": [
        "## ChIP-Histone"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Oq5nwdhCqyVE"
      },
      "source": [
        "Histone modification marks are captured by the `CHIP_HISTONE` output. Here is an\n",
        "example of predictions in the same interval and tissues for histone marks\n",
        "thought to indicate key regulatory effects (e.g., H3K4me3).\n",
        "\n",
        "Some additional considerations for visualizing `CHIP_HISTONE` predictions:\n",
        "\n",
        "*   ChIP-Histone (and ChIP-TF) predictions are returned at a coarser resolution\n",
        "    (128bp), so we need to adjust the interval size plotted to be compatible\n",
        "    (i.e., its width must be a multiple of 128). This is done automatically by\n",
        "    the plotting library.\n",
        "*   Not all biosamples have predictions for all histone marks, but four of the\n",
        "    major ones (H3K4me3, H3K4me1, H3K27ac, H3K27me3 and H3K36me3) are available\n",
        "    for 40% of biosamples. Let's see which histone marks are covered by\n",
        "    biosamples corresponding to colon tissue:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "oplX3UfpucoO"
      },
      "outputs": [],
      "source": [
        "output_metadata.chip_histone[\n",
        "    output_metadata.chip_histone['biosample_name'].str.contains('colon')\n",
        "]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "T8U7yiGMqyVE"
      },
      "outputs": [],
      "source": [
        "# List of IDs corresponding to various colon tissues in `CHIP_HISTONE` output.\n",
        "ontology_terms = [\n",
        "    'UBERON:0000317',\n",
        "    'UBERON:0001155',\n",
        "    'UBERON:0001157',\n",
        "    'UBERON:0001159',\n",
        "]\n",
        "\n",
        "# Make predictions.\n",
        "output = dna_model.predict_interval(\n",
        "    interval=interval,\n",
        "    requested_outputs={dna_client.OutputType.CHIP_HISTONE},\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uY9FpDS4jcWN"
      },
      "source": [
        "We can extract the locations of transcription start sites as annotated by\n",
        "GENCODE using `gene_annotation.extract_tss()`, and plot these using\n",
        "`plot_components.IntervalAnnotation()`:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "SIxvrfPSd3W5"
      },
      "outputs": [],
      "source": [
        "gtf_tss = gene_annotation.extract_tss(gtf_longest_transcript)\n",
        "\n",
        "tss_as_intervals = [\n",
        "    genome.Interval(\n",
        "        chromosome=row.Chromosome,\n",
        "        start=row.Start,\n",
        "        end=row.End + 1000,  # Add extra 1Kb so the TSSs are visible.\n",
        "        name=row.gene_name,\n",
        "    )\n",
        "    for _, row in gtf_tss.iterrows()\n",
        "]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qdastNqEg8Vc"
      },
      "source": [
        "We can also reorder tracks so they are grouped by histone mark, and also specify\n",
        "colors based on their histone mark."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "DKasK1YEg7DB"
      },
      "outputs": [],
      "source": [
        "reordered_chip_histone = output.chip_histone.select_tracks_by_index(\n",
        "    output.chip_histone.metadata.sort_values('histone_mark').index\n",
        ")\n",
        "\n",
        "histone_to_color = {\n",
        "    'H3K27AC': '#e41a1c',\n",
        "    'H3K36ME3': '#ff7f00',\n",
        "    'H3K4ME1': '#377eb8',\n",
        "    'H3K4ME3': '#984ea3',\n",
        "    'H3K9AC': '#4daf4a',\n",
        "    'H3K27ME3': '#ffc0cb',\n",
        "}\n",
        "\n",
        "track_colors = (\n",
        "    reordered_chip_histone.metadata['histone_mark']\n",
        "    .map(lambda x: histone_to_color.get(x.upper(), '#000000'))\n",
        "    .values\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "O_YZ5pwWqyVE"
      },
      "outputs": [],
      "source": [
        "# Build plot.\n",
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=reordered_chip_histone,\n",
        "            ylabel_template=(\n",
        "                'CHIP HISTONE: {biosample_name} ({strand})\\n{histone_mark}'\n",
        "            ),\n",
        "            filled=True,\n",
        "            track_colors=track_colors,\n",
        "        ),\n",
        "    ],\n",
        "    interval=interval,\n",
        "    annotations=[\n",
        "        plot_components.IntervalAnnotation(\n",
        "            tss_as_intervals, alpha=0.5, colors='blue'\n",
        "        )\n",
        "    ],\n",
        "    despine_keep_bottom=True,\n",
        "    title='Predicted histone modification markers in colon tissue',\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P6CGieJNIqf2"
      },
      "source": [
        "We can see that many of the predicted histone modification peaks overlap with\n",
        "transcription start sites, especially markers associated with promoters\n",
        "(H3K4me3, H3K27ac, and H3K9ac). Other markers such as H3K36me3 and H3K4me1 are\n",
        "more associated with gene bodies and enhancers, so these signals are less\n",
        "pronounced around the TSSs."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FHDs_Vq94SpV"
      },
      "source": [
        "## ChIP-TF"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eUr7IcCW4SpV"
      },
      "source": [
        "Transcription factor (TF) binding patterns are captured by the model's `CHIP_TF`\n",
        "output. We will visualize these in a similar way to `CHIP_HISTONE` predictions,\n",
        "with one difference: we need to modify the ontology terms we request in order to\n",
        "get good coverage of both biosamples of interest and multiple TFs. Many\n",
        "biosamples have predictions for two major TFs, namely CTCF and POLR2A, but some\n",
        "cell-lines (HepG2 and K562) have coverage over many more TFs (501 and 269\n",
        "respectively).\n",
        "\n",
        "Here, we will also demonstrate two helpful utilities: selecting specific tracks\n",
        "from a `TrackData` object and aggregating predictions across tracks in a\n",
        "`TrackData` object."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OTOi-Ktb4SpW"
      },
      "outputs": [],
      "source": [
        "ontology_terms = [\n",
        "    'UBERON:0001159',  # Sigmoid colon.\n",
        "    'UBERON:0001157',  # Transverse colon.\n",
        "    'EFO:0002067',  # K562.\n",
        "    'EFO:0001187',  # HepG2.\n",
        "]\n",
        "\n",
        "output = dna_model.predict_interval(\n",
        "    interval=interval,\n",
        "    requested_outputs={dna_client.OutputType.CHIP_TF},\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yMSXBhrDDGu8"
      },
      "source": [
        "Say we then only want to work with the K562 and HepG2 part of the output. We can\n",
        "accomplish this by calling `filter_tracks`:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dx9tlBcQCXzR"
      },
      "outputs": [],
      "source": [
        "ontology_terms = [\n",
        "    'EFO:0002067',  # K562.\n",
        "    'EFO:0001187',  # HepG2.\n",
        "]\n",
        "\n",
        "output_chip_tf = output.chip_tf.filter_tracks(\n",
        "    (output.chip_tf.metadata['ontology_curie'].isin(ontology_terms)).values\n",
        ")\n",
        "len(output_chip_tf.metadata)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5CcAfBOfDdxt"
      },
      "source": [
        "In these 845 tracks across the 2 cell types, the maximum predicted values vary\n",
        "considerably:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4gn11NP_W3mJ"
      },
      "outputs": [],
      "source": [
        "max_predictions = output_chip_tf.metadata[\n",
        "    ['ontology_curie', 'biosample_name', 'transcription_factor']\n",
        "].copy()\n",
        "\n",
        "max_predictions.loc[:, 'max_prediction'] = output_chip_tf.values.max(axis=0)\n",
        "max_predictions.sort_values('max_prediction', ascending=False).reset_index(\n",
        "    drop=True\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5bX1nHzBXnpU"
      },
      "source": [
        "We can filter the tracks down to only those with a maximum value of at least\n",
        "10000:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "uv6jLzhvD64q"
      },
      "outputs": [],
      "source": [
        "print(f'Number of tracks before filtering: {len(output_chip_tf.metadata)}')\n",
        "\n",
        "output_filtered = output_chip_tf.filter_tracks(\n",
        "    output_chip_tf.values.max(axis=0) > 8000\n",
        ")\n",
        "print(f'Number of tracks after filtering: {len(output_filtered.metadata)}')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PgtgnV8OYNoK"
      },
      "source": [
        "This is a more manageable number of tracks to visualize. Let's build up the plot\n",
        "as before:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jaUefFV-4SpW"
      },
      "outputs": [],
      "source": [
        "# Build plot.\n",
        "plot_components.plot(\n",
        "    components=[\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output_filtered,\n",
        "            ylabel_template=(\n",
        "                'CHIP TF: {biosample_name} ({strand})\\n{transcription_factor}'\n",
        "            ),\n",
        "            filled=True,\n",
        "        ),\n",
        "    ],\n",
        "    interval=interval,\n",
        "    title='Predicted TF-binding in K562 and HepG2 cell-lines.',\n",
        "    despine_keep_bottom=True,\n",
        "    annotations=[\n",
        "        plot_components.IntervalAnnotation(\n",
        "            tss_as_intervals, alpha=0.3, colors='blue'\n",
        "        )\n",
        "    ],\n",
        ")\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "q5fvBZnqHp2A"
      },
      "source": [
        "In this figure, we see that CTCF binding sites strongly coincide with those of\n",
        "the cohesin complex protein RAD21. Binding locations for the same protein (e.g.,\n",
        "RBFOX2, REST) appear largely the same across these two cell types, with some\n",
        "change in magnitude. Peaks in predicted POLR2G binding, a component of RNA\n",
        "Polymerase II, typically occur near TSS sites.\n",
        "\n",
        "We can use the same process to select tracks with high peaks anywhere within the\n",
        "*APOL4* gene interval:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "W2GwYd-8Zt61"
      },
      "outputs": [],
      "source": [
        "# Compute max predicted values per track in the APOL4 gene interval.\n",
        "max_predictions = output.chip_tf.slice_by_interval(\n",
        "    apol4_interval, match_resolution=True\n",
        ").values.max(axis=0)\n",
        "\n",
        "# Filter to the 10 tracks with the highest predictions.\n",
        "output_filtered = output.chip_tf.filter_tracks(\n",
        "    (max_predictions >= np.sort(max_predictions)[-10])\n",
        ")\n",
        "\n",
        "# Build plot.\n",
        "plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=output_filtered,\n",
        "            ylabel_template=(\n",
        "                'CHIP TF: {biosample_name} ({strand})\\n{transcription_factor}'\n",
        "            ),\n",
        "            filled=True,\n",
        "        ),\n",
        "    ],\n",
        "    interval=apol4_interval,\n",
        "    annotations=[plot_components.IntervalAnnotation(promoter_intervals)],\n",
        "    despine_keep_bottom=True,\n",
        "    title='Predicted TF-binding in K562, HepG2, and sigmoid colon.',\n",
        ")\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2pkxPpeeEvst"
      },
      "source": [
        "We observe a strong predicted CTCF peak near the second-to-last exon (counting\n",
        "from the right), which is predicted across different tissues. We also see peaks\n",
        "for (phosphorylated) POL2 binding around one of the putative promoters.\n",
        "\n",
        "We can average the signal for a transcription factor across tissues to simplify\n",
        "the plot. For example, to plot the mean CTCF signal across tissues, we first\n",
        "need to construct a new `TrackData` object:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CQwDVKAFhheC"
      },
      "outputs": [],
      "source": [
        "mean_ctcf = output_filtered.values[\n",
        "    :, output_filtered.metadata['transcription_factor'] == 'CTCF'\n",
        "].mean(axis=1)\n",
        "\n",
        "# Construct a new TrackData object from the mean values.\n",
        "tdata_mean_ctcf = track_data.TrackData(\n",
        "    values=mean_ctcf[:, None],\n",
        "    metadata=pd.DataFrame(\n",
        "        {'transcription_factor': ['CTCF'], 'name': ['mean'], 'strand': ['.']}\n",
        "    ),\n",
        "    interval=output_filtered.interval,\n",
        "    resolution=output_filtered.resolution,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NkBP0l8zn41d"
      },
      "source": [
        "And then plot as usual:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pibKAGe1jKhm"
      },
      "outputs": [],
      "source": [
        "plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(transcripts),\n",
        "        plot_components.Tracks(\n",
        "            tdata=tdata_mean_ctcf,\n",
        "            ylabel_template='{name} {transcription_factor}',\n",
        "            filled=True,\n",
        "        ),\n",
        "    ],\n",
        "    interval=apol4_interval,\n",
        "    annotations=[plot_components.IntervalAnnotation(promoter_intervals)],\n",
        "    despine_keep_bottom=True,\n",
        "    title='Predicted CTCF binding (mean across cell types)',\n",
        ")\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XLknUdRRGJPE"
      },
      "source": [
        "## Contact maps"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fc9LMc_KGJPF"
      },
      "source": [
        "Relative frequency of physical contacts between pairwise genetic positions are\n",
        "captured by the model's `CONTACT_MAPS` output. This output type is provided at\n",
        "even coarser resolution (2048bp), so we can only consider intervals that span\n",
        "distances larger than this.\n",
        "\n",
        "Let's make the contact map predictions for a colon cell line:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1KD-Ac1YLjle"
      },
      "outputs": [],
      "source": [
        "ontology_terms = [\n",
        "    'EFO:0002824',  # HCT116 colon carcinoma cell line.\n",
        "]\n",
        "\n",
        "output = dna_model.predict_interval(\n",
        "    interval=interval,\n",
        "    requested_outputs={dna_client.OutputType.CONTACT_MAPS},\n",
        "    ontology_terms=ontology_terms,\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6VS7O_kdkJGB"
      },
      "source": [
        "And plot the predicted DNA-DNA contacts:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "riZLzUZZJFzZ"
      },
      "outputs": [],
      "source": [
        "plot = plot_components.plot(\n",
        "    [\n",
        "        plot_components.TranscriptAnnotation(longest_transcripts),\n",
        "        plot_components.ContactMaps(\n",
        "            tdata=output.contact_maps,\n",
        "            ylabel_template='{biosample_name}\\n{name}',\n",
        "            cmap='autumn_r',\n",
        "            vmax=1.0,\n",
        "        ),\n",
        "    ],\n",
        "    interval=interval,\n",
        "    title='Predicted contact maps',\n",
        ")\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ymPfZ4Qckesf"
      },
      "source": [
        "Three prominent interacting regions, potentially representing\n",
        "topologically-associated domains (TADs), are visible as blocks. This interaction\n",
        "signal appears to be stronger in the first plot."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BnHh7sw2RpAH"
      },
      "source": [
        "## Conclusion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4dTTcGnTCh1I"
      },
      "source": [
        "Congratulations! You have now completed the tour of the visualization library\n",
        "and learned how to visualize the core modalities predicted by AlphaGenome."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}